p2p, refactor: performance improvements to ProtectEvictionCandidatesByRatio()

jonatack opened the pull request:
This follow-up to #21261 improves `ProtectEvictionCandidatesByRatio()` for better performance. 

Benchmarks are added; the performance improvement is between 2x and 5x for the benchmarked cases (CPU 2.50GHz, Turbo off, performance mode, Debian Clang 11 non-debug build).

```
$ ./src/bench/bench_bitcoin -filter="EvictionProtection*.*"
```

The refactored code is well-covered by existing unit tests and also a fuzzer.

- `$ ./src/test/test_bitcoin -t net_peer_eviction_tests`
- `$ FUZZ=node_eviction ./src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/node_eviction`

vasild committed:
p2p: earlier continuation when no remaining eviction candidates

in ProtectEvictionCandidatesByRatio().

With this change, `if (n.count == 0) continue;` will be true
if a network had candidates protected in the first iterations
and has no candidates remaining to be protected in later iterations.

Co-authored-by: Jon Atack <jon@atack.com>

jonatack committed:
p2p: iterate eviction protection only on networks having candidates

in ProtectEvictionCandidatesByRatio().

Thank you to Vasil Dimov, whose suggestions during a post-merge
discussion about PR 21261 reminded me that I had done this in
earlier versions of the PR, e.g. commits like ef411cd2.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack commented:
`./src/test/test_bitcoin -t net_peer_eviction_tests/peer_protection_test` runs about 10% faster for me on this branch, though it is a debug build and the unit tests use very low numbers of eviction candidates (between 4 and 24).  I'll write a test with higher numbers of eviction candidates and benchmark with optimized builds (and fix the fuzzer CI failure).

jonatack commented:
<details><summary>With this 2-line test change</summary><p>

```diff
diff --git a/src/test/net_peer_eviction_tests.cpp b/src/test/net_peer_eviction_tests.cpp
index 4bfd487b86..f6ea4ba1af 100644
--- a/src/test/net_peer_eviction_tests.cpp
+++ b/src/test/net_peer_eviction_tests.cpp
@@ -85,6 +85,7 @@ BOOST_AUTO_TEST_CASE(peer_protection_test)
 {
     FastRandomContext random_context{true};
     int num_peers{12};
+    for (int i = 0; i < 500000; ++i) {
 
     // Expect half of the peers with greatest uptime (the lowest nTimeConnected)
     // to be protected from eviction.
@@ -456,6 +457,7 @@ BOOST_AUTO_TEST_CASE(peer_protection_test)
         /* protected_peer_ids */ {0, 1, 2, 3, 4, 5, 7, 8, 11, 12, 16, 17},
         /* unprotected_peer_ids */ {6, 9, 10, 13, 14, 15, 18, 19, 20, 21, 22, 23},
         random_context));
+    }
 }
```
</p></details>

```
$ time ./src/test/test_bitcoin -t net_peer_eviction_tests/peer_protection_test
```

![Screenshot from 2021-06-19 23-16-38](https://user-images.githubusercontent.com/2415484/122655715-1a0d6980-d144-11eb-80c7-28cc1646369e.jpg)

Seeing roughly an 8% speedup, Clang 11 non-debug build on Debian testing, run times in seconds on this branch vs on master. So pretty much the same as the comment above.

jonatack committed:
bench: add peer eviction protection benchmarks

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack commented:
Added a benchmark. This pull speeds up `ProtectEvictionCandidatesByRatio()` between 2x and 4x for the benchmarked cases. It's not a hotspot, but the improvement is significant.


```
$ ./src/bench/bench_bitcoin -filter=EvictionProtection*.*
```

master

```rake
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    5,403,043,444.00 |                0.19 |    1.8% |     59.74 | `EvictionProtection1Network250Candidates`
|    2,946,547,379.00 |                0.34 |    2.7% |     33.45 | `EvictionProtection3Networks050Candidates`
|    2,798,691,071.00 |                0.36 |    2.2% |     30.79 | `EvictionProtection3Networks100Candidates`
|    7,288,364,370.00 |                0.14 |    0.7% |     80.16 | `EvictionProtection3Networks250Candidates`
```

this PR

```rake
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    1,183,198,205.00 |                0.85 |    1.2% |     13.12 | `EvictionProtection1Network250Candidates`
|    1,151,709,186.00 |                0.87 |    1.7% |     12.67 | `EvictionProtection3Networks050Candidates`
|    1,261,170,266.00 |                0.79 |    2.9% |     14.26 | `EvictionProtection3Networks100Candidates`
|    1,157,625,808.00 |                0.86 |    1.0% |     12.95 | `EvictionProtection3Networks250Candidates`
```

------

(intermediate commits)

*p2p: iterate over eviction candidates once instead of thrice*

```rake
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    3,952,529,089.00 |                0.25 |    1.8% |     43.47 | `EvictionProtection1Network250Candidates`
|    2,239,435,919.00 |                0.45 |    3.8% |     25.85 | `EvictionProtection3Networks050Candidates`
|    2,060,389,113.00 |                0.49 |    1.9% |     22.85 | `EvictionProtection3Networks100Candidates`
|    5,258,681,959.00 |                0.19 |    1.0% |     58.26 | `EvictionProtection3Networks250Candidates`
```

*p2p: iterate eviction protection only on networks having candidates*

```rake
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    1,224,597,989.00 |                0.82 |    2.2% |     13.60 | `EvictionProtection1Network250Candidates`
```

klementtan commented:
Tested on macOs 11.4 and this PR indeed optimize it by 1.3x to 5x. 

<details> 
<summary> Test result </summary>

On 272f327fbfcb8bb61f8e1c9d65c665372cd60976(master with benchmarks):
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      756,331,588.00 |                1.32 |    1.2% |      8.38 | `EvictionProtection1Network250Candidates`
|    1,579,971,776.00 |                0.63 |    1.1% |     17.45 | `EvictionProtection3Networks050Candidates`
|    1,395,431,168.00 |                0.72 |    1.0% |     15.70 | `EvictionProtection3Networks100Candidates`
|    3,716,091,921.00 |                0.27 |    1.0% |     41.32 | `EvictionProtection3Networks250Candidates`
```
This PR:
```          
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      492,425,456.00 |                2.03 |    0.9% |      5.45 | `EvictionProtection1Network250Candidates`
|      739,073,603.00 |                1.35 |    1.3% |      8.13 | `EvictionProtection3Networks050Candidates`
|      937,281,186.00 |                1.07 |    0.7% |     10.38 | `EvictionProtection3Networks100Candidates`
|      828,820,224.00 |                1.21 |    1.6% |      9.10 | `EvictionProtection3Networks250Candidates`
```
</details>

Not sure if it is an issue on my side but I had to test with `./src/bench/bench_bitcoin -filter="EvictionProtection*.*"`(quotes around regex) instead of the command provided by the PR description.

jarolrod reviewed (commented):
ACK ab2993336fbdff72d125f906c2baec4c5dcb79e3

This is quite a performance improvement and is a clear change which is easy to follow. My benchmark tests confirm the performance improvement. Thanks for working on this improvement and for adding additional benchmark coverage.

Ran benchmarks upon each intermediate commit to test performance improvement at every stage. Cherry-picked each commit on top of current master and ran unit tests which each commit cherry-picked.

**Master**
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    2,283,211,027.00 |                0.44 |    0.1% |     25.13 | `EvictionProtection1Network250Candidates`
|    1,417,424,933.00 |                0.71 |    0.1% |     15.59 | `EvictionProtection3Networks050Candidates`
|    1,293,719,559.00 |                0.77 |    0.1% |     14.25 | `EvictionProtection3Networks100Candidates`
|    3,350,000,724.00 |                0.30 |    0.1% |     36.88 | `EvictionProtection3Networks250Candidates`
```

**PR**
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      778,011,162.00 |                1.29 |    0.3% |      8.56 | `EvictionProtection1Network250Candidates`
|      811,555,934.00 |                1.23 |    0.1% |      8.93 | `EvictionProtection3Networks050Candidates`
|      806,194,432.00 |                1.24 |    0.1% |      8.87 | `EvictionProtection3Networks100Candidates`
|      776,191,251.00 |                1.29 |    0.1% |      8.54 | `EvictionProtection3Networks250Candidates`
```

### Notes on commits 

- [x] 272f327fbfcb8bb61f8e1c9d65c665372cd60976
  - Definition of test is appropriate and effective in terms of benchmarking what we care about
- [x] 99a25fb393c791bb6b6edeab3677d39bf6b1d97d
  - the performance boost is from moving away from looping through each network and then interating through all `eviction_candidates` -> iterate through `eviction_candidates` once; for each candidate we iterate through add count to appropriate network
  - Still appropriately counts eviction candidates per network
  - Performance improvement at this stage is very modest compared to `master+272f327fbfcb8bb61f8e1c9d65c665372cd60976`
- [x] 06b898c3e126ff7f1a17d8d3a5f7c9e1aad0ea06
  - uses lambda expression in `std::count_if` to efficiently determine networks we have candidates from
  - PR description states that "... increase the number protected per peer iteration", this would only come into play if there was a descrease in the number of networks to protect because we detected there were no candidates in that network. If you have to retouch, can you document this more clearly?
  - this and 99a25fb393c791bb6b6edeab3677d39bf6b1d97d make up the bulk of the performance improvement
- [x] 478534340e9b30f4e906da290e3b763dec6c945c 
  - makes sense 🥃
- [x] ab2993336fbdff72d125f906c2baec4c5dcb79e3
  - Is an additional performance improvement

theStack commented:
Concept ACK

jonatack commented:
Thanks for the review @jarolrod!  If anyone else would like to review or ACK that would be great as this completes merged PR #21261.

vasild commented on src/bench/peer_eviction.cpp:
```diff
@@ -0,0 +1,169 @@
+#include <functional>
+#include <vector>
+
+std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(int num_peers, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    for (int id = 0; id < num_peers; ++id) {
+        candidates.push_back({
+            /* id */ id,
+            /* nTimeConnected */ static_cast<int64_t>(random_context.randrange(100)),
+            /* m_min_ping_time */ std::chrono::microseconds{random_context.randrange(100)},
+            /* nLastBlockTime */ static_cast<int64_t>(random_context.randrange(100)),
+            /* nLastTXTime */ static_cast<int64_t>(random_context.randrange(100)),
+            /* fRelevantServices */ random_context.randbool(),
+            /* fRelayTxes */ random_context.randbool(),
+            /* fBloomFilter */ random_context.randbool(),
+            /* nKeyedNetGroup */ random_context.randrange(100),
+            /* prefer_evict */ random_context.randbool(),
+            /* m_is_local */ random_context.randbool(),
+            /* m_network */ ALL_NETWORKS[random_context.randrange(ALL_NETWORKS.size())],
+        });
+    }
+    return candidates;
+}
```
This function already exists in `src/test/net_peer_eviction_tests.cpp`. Dedup:

```diff
    test: make GetRandomNodeEvictionCandidates() reusable

    Move `GetRandomNodeEvictionCandidates()` from
    `src/test/net_peer_eviction_tests.cpp` to `src/test/util/net.h` and
    `src/test/util/net.cpp` so that it can be reused by other tests.

diff --git a/src/test/net_peer_eviction_tests.cpp b/src/test/net_peer_eviction_tests.cpp
index 4bfd487b86..5eb280b498 100644
--- a/src/test/net_peer_eviction_tests.cpp
+++ b/src/test/net_peer_eviction_tests.cpp
@@ -14,34 +14,12 @@
 #include <optional>
 #include <unordered_set>
 #include <vector>

 BOOST_FIXTURE_TEST_SUITE(net_peer_eviction_tests, BasicTestingSetup)

-std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_candidates, FastRandomContext& random_context)
-{
-    std::vector<NodeEvictionCandidate> candidates;
-    for (int id = 0; id < n_candidates; ++id) {
-        candidates.push_back({
-            /* id */ id,
-            /* nTimeConnected */ static_cast<int64_t>(random_context.randrange(100)),
-            /* m_min_ping_time */ std::chrono::microseconds{random_context.randrange(100)},
-            /* nLastBlockTime */ static_cast<int64_t>(random_context.randrange(100)),
-            /* nLastTXTime */ static_cast<int64_t>(random_context.randrange(100)),
-            /* fRelevantServices */ random_context.randbool(),
-            /* fRelayTxes */ random_context.randbool(),
-            /* fBloomFilter */ random_context.randbool(),
-            /* nKeyedNetGroup */ random_context.randrange(100),
-            /* prefer_evict */ random_context.randbool(),
-            /* m_is_local */ random_context.randbool(),
-            /* m_network */ ALL_NETWORKS[random_context.randrange(ALL_NETWORKS.size())],
-        });
-    }
-    return candidates;
-}
-
 // Create `num_peers` random nodes, apply setup function `candidate_setup_fn`,
 // call ProtectEvictionCandidatesByRatio() to apply protection logic, and then
 // return true if all of `protected_peer_ids` and none of `unprotected_peer_ids`
 // are protected from eviction, i.e. removed from the eviction candidates.
 bool IsProtected(int num_peers,
                  std::function<void(NodeEvictionCandidate&)> candidate_setup_fn,
diff --git a/src/test/util/net.cpp b/src/test/util/net.cpp
index 847a490e03..64b70a61f0 100644
--- a/src/test/util/net.cpp
+++ b/src/test/util/net.cpp
@@ -34,6 +34,28 @@ bool ConnmanTestMsg::ReceiveMsgFrom(CNode& node, CSerializedNetMsg& ser_msg) con

     bool complete;
     NodeReceiveMsgBytes(node, ser_msg_header, complete);
     NodeReceiveMsgBytes(node, ser_msg.data, complete);
     return complete;
 }
+
+std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_candidates, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    for (int id = 0; id < n_candidates; ++id) {
+        candidates.push_back({
+            /* id */ id,
+            /* nTimeConnected */ static_cast<int64_t>(random_context.randrange(100)),
+            /* m_min_ping_time */ std::chrono::microseconds{random_context.randrange(100)},
+            /* nLastBlockTime */ static_cast<int64_t>(random_context.randrange(100)),
+            /* nLastTXTime */ static_cast<int64_t>(random_context.randrange(100)),
+            /* fRelevantServices */ random_context.randbool(),
+            /* fRelayTxes */ random_context.randbool(),
+            /* fBloomFilter */ random_context.randbool(),
+            /* nKeyedNetGroup */ random_context.randrange(100),
+            /* prefer_evict */ random_context.randbool(),
+            /* m_is_local */ random_context.randbool(),
+            /* m_network */ ALL_NETWORKS[random_context.randrange(ALL_NETWORKS.size())],
+        });
+    }
+    return candidates;
+}
diff --git a/src/test/util/net.h b/src/test/util/net.h
index 1b49a671bd..66f6cc61b8 100644
--- a/src/test/util/net.h
+++ b/src/test/util/net.h
@@ -138,7 +138,9 @@ public:

 private:
     const std::string m_contents;
     mutable size_t m_consumed;
 };

+std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_candidates, FastRandomContext& random_context);
+
 #endif // BITCOIN_TEST_UTIL_NET_H
```

jonatack replied:
Done (thanks!)

vasild commented on src/bench/peer_eviction.cpp:
```diff
@@ -0,0 +1,169 @@
+            },
+            random_context)};
+
+    bench.run([&] {
+        std::vector<NodeEvictionCandidate> candidates;
+        for (int id = 0; id < 60000; ++id) {
+            candidates = eviction_candidates;
+            ProtectEvictionCandidatesByRatio(candidates);
+        }
+    });
```
I think [nanobench](https://nanobench.ankerl.com/reference.html#_CPPv4N6ankerl9nanobench5Bench6epochsE6size_t) is designed to make such loops unnecessary. It can provide stable results even if the test function is very fast/short, no need to loop 60k times.

Also, the above benchmarks the copying of the data which could add noise to the test. Given that the input is not huge we can create a separate copy for each iteration of the test, this consumes a few tens of megabytes which is ok. Here is an example (that also uses the de-duplicated `GetRandomNodeEvictionCandidates()` from `test/util/net.h`):

```cpp
// Copyright (c) 2021 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <bench/bench.h>
#include <net.h>
#include <netaddress.h>
#include <random.h>
#include <test/util/net.h>
#include <test/util/setup_common.h>

#include <algorithm>
#include <functional>
#include <vector>

static void EvictionProtectionCommon(
    benchmark::Bench& bench,
    int num_candidates,
    std::function<void(NodeEvictionCandidate&)> candidate_setup_fn,
    size_t epochs,
    uint64_t epoch_iterations)
{
    using Candidates = std::vector<NodeEvictionCandidate>;

    FastRandomContext random_context{true};

    bench.epochs(epochs).epochIterations(epoch_iterations);

    Candidates candidates{GetRandomNodeEvictionCandidates(num_candidates, random_context)};
    for (auto& c : candidates) {
        candidate_setup_fn(c);
    }

    std::vector<Candidates> copies{bench.epochs() * bench.epochIterations(), candidates};
    size_t i{0};
    bench.run([&] {
        ProtectEvictionCandidatesByRatio(copies.at(i));
        ++i;
    });
}

/* Benchmarks */

static void EvictionProtection1Network250Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        250 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = false;
            if (c.id >= 130 && c.id < 240) { // 110 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV6;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

static void EvictionProtection3Networks050Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        50 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = (c.id == 28 || c.id == 47); //  2 localhost
            if (c.id >= 30 && c.id < 47) {             // 17 I2P
                c.m_network = NET_I2P;
            } else if (c.id >= 24 && c.id < 28) { //  4 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV4;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

static void EvictionProtection3Networks100Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        100 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = (c.id >= 55 && c.id < 60); //  5 localhost
            if (c.id >= 70 && c.id < 80) {            // 10 I2P
                c.m_network = NET_I2P;
            } else if (c.id >= 80 && c.id < 96) { // 16 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV4;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

static void EvictionProtection3Networks250Candidates(benchmark::Bench& bench)
{
    EvictionProtectionCommon(
        bench,
        250 /* num_candidates */,
        [](NodeEvictionCandidate& c) {
            c.nTimeConnected = c.id;
            c.m_is_local = (c.id >= 140 && c.id < 160); // 20 localhost
            if (c.id >= 170 && c.id < 180) {            // 10 I2P
                c.m_network = NET_I2P;
            } else if (c.id >= 190 && c.id < 240) { // 50 Tor
                c.m_network = NET_ONION;
            } else {
                c.m_network = NET_IPV4;
            }
        },
        10 /* epochs */,
        100 /* epoch_iterations */);
}

// 1 disadvantaged network (Tor) with 250 eviction candidates.
BENCHMARK(EvictionProtection1Network250Candidates);

// 3 disadvantaged networks (I2P/localhost/Tor) with 50/100/250 eviction candidates.
BENCHMARK(EvictionProtection3Networks050Candidates);
BENCHMARK(EvictionProtection3Networks100Candidates);
BENCHMARK(EvictionProtection3Networks250Candidates);
```

jonatack replied:
I tried this, and am playing with your proposal that looks good! but as before when I was trying to use these options, nanobench kept saying it was unstable and to increase the iterations, beyond even the numbers I settled on in my manual loops that yielded slowish but stableish results for me.
```
|            9,132.53 |          109,498.74 |    9.8% |      0.10 | :wavy_dash: `EvictionProtection3Networks050Candidates` (Unstable with ~1,000.0 iters. Increase `minEpochIterations` to e.g. 10000)
|           21,737.81 |           46,002.80 |    6.8% |      0.22 | :wavy_dash: `EvictionProtection3Networks100Candidates` (Unstable with ~1,000.0 iters. Increase `minEpochIterations` to e.g. 10000)
```
So I'm still looking at this and may punt on using the options.  

Yes, I benched the data copying alone, and it seemed insignificant relative to the call under test, not worth messing with by making (potentially many, e.g. 100k) copies instead.  That could be ok if we don't need so many copies, so I understand why you proposed reducing the number of iters.

vasild replied:
Here are some numbers:

* `sizeof(NodeEvictionCandidate)` is 64
* number of candidates is 250
* the current test contains 60k iterations

So that would make 64 * 250 * 60000 = 916 MiB.

In the changes above I suggested 10 epochs * 100 iterations (1000 copies): 64 * 250 * 1000 = 15 MiB.

> so I understand why you proposed reducing the number of iters

Yes, I was concerned about the memory usage, but also I stopped getting the "unstable results" warning at around 500 iters so I decided to double it to 1000 just in case, since 15 MiB is not too much. I guess a few times more than that would also be ok.

vasild replied:
Two random thoughts:

1. Do you get "unstable results" warnings when running other benchmarks from `src/bench`?
2. Nanobench's [`warmup()`](https://nanobench.ankerl.com/reference.html#_CPPv4N6ankerl9nanobench5Bench6warmupE8uint64_t) may help?

jonatack replied:
I've tuned my system for benchmarking and any iterations less than in my original proposal gives errors above 5%, often 10-25%, rendering the benchmark useless for me.  I'm going to consider that my system is unusually slow and unsuited to running these and use this feedback provided that reviewers find the results stable for them.

jonatack replied:
I've tuned my system for benchmarking, and nevertheless, any iterations less than in my original proposal still return bench error rates above 5%, often 10-25%, rendering the benchmark somewhat useless for me.

I'm going to consider that my system (2 physical cores, 4 logical ones, with an open source coreboot) is unusual and unsuited to running these benchmarks and use this feedback provided that reviewers find the results stable for them.

vasild commented on src/net.cpp:
```diff
@@ -921,11 +921,13 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
-                                [&n](const NodeEvictionCandidate& c) {
-                                    return n.is_local ? c.m_is_local : c.m_network == n.id;
-                                });
+    for (auto& c : eviction_candidates) {
+        for (int i{networks.size() - 1}; i >= 0; --i) {
+            if (networks[i].is_local ? c.m_is_local : c.m_network == networks[i].id) {
+                ++networks[i].count;
+                break;
+            }
+        }
     }
```
99a25fb393 `p2p: iterate over eviction candidates once instead of thrice`

I did not observe any difference in performance due to this commit.

It changes the complexity from `O(num networks * num candidates)` to `O(num candidates * num networks)`. The early `break` optimization maybe makes it `O(num candidates * num networks * 0.5)` on average, but that is equal to `O(num candidates * num networks)`.

Just to experiment I made this `O(num candidates)` which is as fast as it could be (unfortunately a bit hackish due to the hardcoded indexes):

```diff
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -918,18 +918,26 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     // array members have first opportunity to recover unused slots from the previous iteration.
     struct Net { bool is_local; Network id; size_t count; };
     std::array<Net, 3> networks{
         {{false, NET_I2P, 0}, {/* localhost */ true, NET_MAX, 0}, {false, NET_ONION, 0}}};
 
     // Count and store the number of eviction candidates per network.
-    for (auto& c : eviction_candidates) {
-        for (int i{networks.size() - 1}; i >= 0; --i) {
-            if (networks[i].is_local ? c.m_is_local : c.m_network == networks[i].id) {
-                ++networks[i].count;
-                break;
-            }
+    for (const auto& c : eviction_candidates) {
+        const Network net = c.m_is_local ? NET_MAX : c.m_network;
+        switch (net) {
+        case NET_I2P:
+            ++networks[0].count;
+            break;
+        case NET_MAX:
+            ++networks[1].count;
+            break;
+        case NET_ONION:
+            ++networks[2].count;
+            break;
+        default:
+            break;
         }
     }
     // Sort `networks` by ascending candidate count, to give networks having fewer candidates
     // the first opportunity to recover unused protected slots from the previous iteration.
```

No difference in performance either. I guess this is operating on too small amount of data to make any performance impact and so this commit can be dropped if its purpose is to improve performance and we can't observe any changes.

vasild replied:
> I did not observe any difference in performance due to this commit.

Here are the results I get:

272f327fbf `bench: add peer eviction protection benchmarks`
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      706,673,971.00 |                1.42 |    1.7% |      7.79 | `EvictionProtection1Network250Candidates`
|    1,299,063,388.00 |                0.77 |    0.7% |     14.30 | `EvictionProtection3Networks050Candidates`
|    1,155,171,747.00 |                0.87 |    0.5% |     12.73 | `EvictionProtection3Networks100Candidates`
|    2,900,919,398.00 |                0.34 |    1.3% |     31.93 | `EvictionProtection3Networks250Candidates`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      714,684,192.00 |                1.40 |    1.0% |      7.91 | `EvictionProtection1Network250Candidates`
|    1,296,579,234.00 |                0.77 |    0.5% |     14.22 | `EvictionProtection3Networks050Candidates`
|    1,159,393,440.00 |                0.86 |    0.7% |     12.71 | `EvictionProtection3Networks100Candidates`
|    2,854,527,930.00 |                0.35 |    0.7% |     31.44 | `EvictionProtection3Networks250Candidates`
```

99a25fb393 `p2p: iterate over eviction candidates once instead of thrice`
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      709,622,212.00 |                1.41 |    1.0% |      7.79 | `EvictionProtection1Network250Candidates`
|    1,310,376,054.00 |                0.76 |    1.2% |     14.39 | `EvictionProtection3Networks050Candidates`
|    1,160,800,660.00 |                0.86 |    0.4% |     12.72 | `EvictionProtection3Networks100Candidates`
|    2,983,258,845.00 |                0.34 |    0.5% |     32.71 | `EvictionProtection3Networks250Candidates`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      711,091,374.00 |                1.41 |    0.5% |      7.82 | `EvictionProtection1Network250Candidates`
|    1,318,781,548.00 |                0.76 |    0.3% |     14.38 | `EvictionProtection3Networks050Candidates`
|    1,142,837,118.00 |                0.88 |    0.2% |     12.40 | `EvictionProtection3Networks100Candidates`
|    2,943,154,182.00 |                0.34 |    0.4% |     32.22 | `EvictionProtection3Networks250Candidates`
```

jonatack replied:
99a25fb393c791bb6b6edeab3677d39bf6b1d97d Yes, ISTM only the first test would utilize the early break and (potentially) show any difference. I agree that it's probably not worth it.

jonatack replied:
99a25fb393c791bb6b6edeab3677d39bf6b1d97d I agree that it's probably not worth it.

jonatack replied:
Dropped.

vasild commented on src/net.cpp:
```diff
@@ -936,27 +938,34 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     size_t num_protected{0};
 
     while (num_protected < max_protect_by_network) {
+        // Count the number of disadvantaged networks from which we have peers to protect.
+        auto num_networks = std::count_if(networks.begin(), networks.end(), [](const Net& n) { return n.count; });
+        if (num_networks == 0) {
+            break;
+        }
```
06b898c3e1 `p2p: iterate eviction protection only on networks having candidates`

Looking at this commit - there is no need to re-calculate `num_networks` inside the loop because `networks` is not changed in the loop. It can be calculated just once before the loop:

```cpp
auto num_networks = std::count_if(networks.begin(), networks.end(), [](const Net& n) { return n.count; });
if (networks > 0) {
    while (num_protected < max_protect_by_network) {
        ...
```

I did not observe any performance change due to this commit.

But considering the subsequent ab2993336f `p2p: earlier continuation when no remaining eviction candidates` (this one brings pefromance boost!) it should be inside the loop because then we change `networks[].count` in the loop. Maybe squash both commits into one?

jonatack replied:
Reorganized to be step by step, with no anticipatory changes.

jonatack committed:
refactor: move GetRandomNodeEvictionCandidates() to test utilities

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

jonatack committed:
p2p: process more candidates per protection iteration

for the usual case when some of the protected networks
don't have eviction candidates, to reduce the number
of iterations in ProtectEvictionCandidatesByRatio().

Picks up an idea in ef411cd2 that I had dropped.

jonatack commented:
Thank you @vasild, @jarolrod, @klementtan and @theStack; feedback taken (and rebased, apologies).

klementtan commented:
Tested on 524aa5dc6a6f20f2fd16a63bd1356d577794d510
- Didn't notice any significant improvement with c0d658f1806a81ae5bf20355861979eb7f7ea1f4
- d2bb6231aaa798590b85700e03dac79f32204aec reduced time for `EvictionProtection1Network250Candidates` by about **40%**
- Didn't notice any significant improvement with edc59d08fc57f5f10657a044224a0b3ae2d2b9bc
- 524aa5dc6a6f20f2fd16a63bd1356d577794d510 reduced time for `EvictionProtection3Networks050Candidates` by about **50%**, `EvictionProtection3Networks100Candidates` by about **40%** and `EvictionProtection3Networks250Candidates` by about **75%**
- Not sure if it is intended but the warning message `:wavy_dash: EvictionProtection3Networks100Candidates (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)` seems to be quite flaky

<details> 
<summary> Bench Results </summary>


**Note:** Results are in reverse chronological order of commits

<details>
<summary>Master + Bench(edebbe7375de02df544ba9c2b566d62b435c0ab1) results </summary>

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           14,383.08 |           69,526.14 |    5.1% |      0.02 | :wavy_dash: `EvictionProtection1Network250Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
|            5,818.85 |          171,855.12 |    0.3% |      0.01 | `EvictionProtection3Networks050Candidates`
|           13,127.23 |           76,177.56 |    1.8% |      0.01 | `EvictionProtection3Networks100Candidates`
|           92,835.73 |           10,771.71 |    7.0% |      0.09 | :wavy_dash: `EvictionProtection3Networks250Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
```

</details>


<details>
<summary>c0d658f1806a81ae5bf20355861979eb7f7ea1f4 results </summary>

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           13,408.72 |           74,578.36 |    1.7% |      0.01 | `EvictionProtection1Network250Candidates`
|            5,610.19 |          178,247.08 |    0.1% |      0.01 | `EvictionProtection3Networks050Candidates`
|           13,536.82 |           73,872.57 |    6.3% |      0.01 | :wavy_dash: `EvictionProtection3Networks100Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
|           87,720.01 |           11,399.91 |    2.2% |      0.09 | `EvictionProtection3Networks250Candidates`
```

</details>

<details>
<summary>d2bb6231aaa798590b85700e03dac79f32204aec results </summary>

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            8,080.88 |          123,748.98 |    2.8% |      0.01 | `EvictionProtection1Network250Candidates`
|            5,650.52 |          176,974.71 |    0.1% |      0.01 | `EvictionProtection3Networks050Candidates`
|           13,847.64 |           72,214.45 |   11.6% |      0.01 | :wavy_dash: `EvictionProtection3Networks100Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
|           88,185.71 |           11,339.71 |    1.8% |      0.09 | `EvictionProtection3Networks250Candidates`
```
</details>


<details>
<summary> edc59d08fc57f5f10657a044224a0b3ae2d2b9bc resutls</summary>
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            7,874.32 |          126,995.09 |    0.9% |      0.01 | `EvictionProtection1Network250Candidates`
|            5,736.20 |          174,331.44 |    1.0% |      0.01 | `EvictionProtection3Networks050Candidates`
|           13,836.98 |           72,270.11 |    9.0% |      0.02 | :wavy_dash: `EvictionProtection3Networks100Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
|           98,370.21 |           10,165.68 |    5.4% |      0.10 | :wavy_dash: `EvictionProtection3Networks250Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
```
</details>


<details>
<summary> 524aa5dc6a6f20f2fd16a63bd1356d577794d510 results </summary>

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            7,854.82 |          127,310.36 |    0.6% |      0.01 | `EvictionProtection1Network250Candidates`
|            2,774.58 |          360,414.26 |    0.1% |      0.00 | `EvictionProtection3Networks050Candidates`
|            8,513.58 |          117,459.47 |    0.1% |      0.01 | `EvictionProtection3Networks100Candidates`
|           22,571.88 |           44,302.90 |   10.1% |      0.02 | :wavy_dash: `EvictionProtection3Networks250Candidates` (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)
```
</details>
</details>

jonatack commented:
Thanks for testing @klementtan!

> Didn't notice any significant improvement with [c0d658f](https://github.com/bitcoin/bitcoin/commit/c0d658f1806a81ae5bf20355861979eb7f7ea1f4)

Thanks for pointing this out. The benchmarks don't test it. Will look at adding a bench with no protected networks.

> Didn't notice any significant improvement with [edc59d0](https://github.com/bitcoin/bitcoin/commit/edc59d08fc57f5f10657a044224a0b3ae2d2b9bc)

This commit doesn't help performance but fixes an off-by-one in the logic that doesn't affect the protection currently, but without it, the last commit would fail the `node_eviction` fuzzer by trying to save (1 - 2) = -1 into `Net::count`, which is `size_t` (unsigned long).

> Not sure if it is intended but the warning message `:wavy_dash: EvictionProtection3Networks100Candidates (Unstable with ~100.0 iters. Increase `minEpochIterations` to e.g. 1000)` seems to be quite flaky

Thanks for reporting.  The benchmark was sped up in the last push, but am looking at maybe increasing the iterations to improve this.

jonatack commented:
Updated per `git diff 524aa5d 4bef482`
- adjusted the benchmarks to use the default `Epochs` of 11, increased the `epochIterations` from 100 to 1000, and added a warmup of 100; I didn't find a magic sweet spot (I have to run it several times) but this seems to improve the `err%` somewhat while keeping the benchmark fast
- added two benchmarks, EvictionProtection0Networks250Candidates and EvictionProtection2Networks250Candidates
- dropped commit https://github.com/bitcoin/bitcoin/commit/edc59d08fc57f5f10657a044224a0b3ae2d2b9bc "p2p: don't protect more than the network candidates count"

<details><summary>benchmark results, CPU at 2.50GHz, Turbo off, non-debug build, most stable runs in <code>err%</code></summary><p>

```
47b8eec bench: add peer eviction protection benchmarks

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           74,657.56 |           13,394.49 |    2.5% |      0.85 | `EvictionProtection0Networks250Candidates`
|           71,502.14 |           13,985.60 |    1.1% |      0.78 | `EvictionProtection1Networks250Candidates`
|           74,242.65 |           13,469.35 |    2.1% |      0.88 | `EvictionProtection2Networks250Candidates`
|            9,280.36 |          107,754.43 |    3.0% |      0.11 | `EvictionProtection3Networks050Candidates`
|           22,320.52 |           44,801.83 |    2.5% |      0.25 | `EvictionProtection3Networks100Candidates`
|          149,821.48 |            6,674.61 |    2.7% |      1.64 | `EvictionProtection3Networks250Candidates`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           73,715.17 |           13,565.73 |    1.9% |      0.82 | `EvictionProtection0Networks250Candidates`
|           71,505.03 |           13,985.03 |    1.7% |      0.79 | `EvictionProtection1Networks250Candidates`
|           74,924.25 |           13,346.81 |    2.6% |      0.86 | `EvictionProtection2Networks250Candidates`
|            9,493.08 |          105,339.84 |    4.4% |      0.11 | `EvictionProtection3Networks050Candidates`
|           23,528.00 |           42,502.55 |    2.7% |      0.26 | `EvictionProtection3Networks100Candidates`
|          148,696.37 |            6,725.11 |    0.6% |      1.68 | `EvictionProtection3Networks250Candidates`



4bef482 p2p: earlier continuation when no remaining eviction candidates

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           19,051.62 |           52,488.98 |    2.6% |      0.21 | `EvictionProtection0Networks250Candidates`
|           25,607.00 |           39,051.82 |    1.9% |      0.28 | `EvictionProtection1Networks250Candidates`
|           31,392.03 |           31,855.22 |    3.5% |      0.40 | `EvictionProtection2Networks250Candidates`
|            4,873.49 |          205,191.84 |    2.1% |      0.06 | `EvictionProtection3Networks050Candidates`
|           13,827.88 |           72,317.64 |    3.9% |      0.16 | `EvictionProtection3Networks100Candidates`
|           34,605.17 |           28,897.42 |    1.5% |      0.39 | `EvictionProtection3Networks250Candidates`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           19,255.11 |           51,934.26 |    2.9% |      0.21 | `EvictionProtection0Networks250Candidates`
|           26,797.82 |           37,316.47 |    3.4% |      0.30 | `EvictionProtection1Networks250Candidates`
|           31,007.27 |           32,250.50 |    1.8% |      0.35 | `EvictionProtection2Networks250Candidates`
|            5,075.53 |          197,023.62 |    3.3% |      0.06 | `EvictionProtection3Networks050Candidates`
|           13,964.15 |           71,611.94 |    3.5% |      0.16 | `EvictionProtection3Networks100Candidates`
|           34,294.63 |           29,159.08 |    1.6% |      0.39 | `EvictionProtection3Networks250Candidates`
```
</p></details>

klementtan commented:
Tested and code review ACK b1d905c2. 

Also no longer seeing `... Unstable with ~100.0 iters. Increase minEpochIterations to e.g. 1000)` warning. 

<details>
<summary> Bench results </summary>

5adb0645:

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           18,612.89 |           53,726.22 |    1.3% |      0.21 | `EvictionProtection0Networks250Candidates`
|           15,604.12 |           64,085.64 |    1.1% |      0.17 | `EvictionProtection1Networks250Candidates`
|           46,098.96 |           21,692.46 |    1.4% |      0.51 | `EvictionProtection2Networks250Candidates`
|            7,573.28 |          132,043.17 |    2.2% |      0.09 | `EvictionProtection3Networks050Candidates`
|           16,631.50 |           60,126.87 |    3.2% |      0.18 | `EvictionProtection3Networks100Candidates`
|           97,775.81 |           10,227.48 |    3.3% |      1.09 | `EvictionProtection3Networks250Candidates`
```

02e411ec
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           18,389.64 |           54,378.44 |    1.7% |      0.20 | `EvictionProtection0Networks250Candidates`
|           13,869.52 |           72,100.52 |    1.3% |      0.15 | `EvictionProtection1Networks250Candidates`
|           46,016.06 |           21,731.55 |    3.4% |      0.50 | `EvictionProtection2Networks250Candidates`
|            6,408.73 |          156,037.13 |    3.6% |      0.07 | `EvictionProtection3Networks050Candidates`
|           16,160.17 |           61,880.55 |    2.1% |      0.18 | `EvictionProtection3Networks100Candidates`
|           96,984.48 |           10,310.93 |    4.5% |      1.07 | `EvictionProtection3Networks250Candidates`
```

c9e8d8f9
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           13,422.68 |           74,500.80 |    1.3% |      0.15 | `EvictionProtection0Networks250Candidates`
|           10,170.63 |           98,322.35 |    1.5% |      0.11 | `EvictionProtection1Networks250Candidates`
|           17,204.80 |           58,123.30 |    1.0% |      0.19 | `EvictionProtection2Networks250Candidates`
|            7,508.82 |          133,176.80 |    2.6% |      0.08 | `EvictionProtection3Networks050Candidates`
|           16,172.34 |           61,833.97 |    1.1% |      0.17 | `EvictionProtection3Networks100Candidates`
|           95,083.59 |           10,517.06 |    4.0% |      1.05 | `EvictionProtection3Networks250Candidates`

```
b1d905c2
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           23,721.76 |           42,155.39 |    1.2% |      0.27 | `EvictionProtection0Networks250Candidates`
|           17,557.76 |           56,954.87 |    0.2% |      0.20 | `EvictionProtection1Networks250Candidates`
|           29,148.45 |           34,307.14 |    0.4% |      0.33 | `EvictionProtection2Networks250Candidates`
|            6,159.56 |          162,349.31 |    0.1% |      0.07 | `EvictionProtection3Networks050Candidates`
|           19,989.47 |           50,026.33 |    0.6% |      0.22 | `EvictionProtection3Networks100Candidates`
|           42,009.48 |           23,804.15 |    0.8% |      0.47 | `EvictionProtection3Networks250Candidates`
```
</details>

vasild reviewed (approved):
ACK b1d905c225e87a4a289c0cd3593c6c21cea3fba7

base (5adb06457403f8c1d874e9c6748ecbb78ef8fa2b bench: add peer eviction protection benchmarks)
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           15,636.87 |           63,951.43 |    0.6% |      0.17 | `EvictionProtection0Networks250Candidates`
|           12,885.39 |           77,607.27 |    0.5% |      0.14 | `EvictionProtection1Networks250Candidates`
|           35,732.17 |           27,985.99 |    0.3% |      0.39 | `EvictionProtection2Networks250Candidates`
|            5,514.20 |          181,349.88 |    1.9% |      0.06 | `EvictionProtection3Networks050Candidates`
|           12,243.04 |           81,679.08 |    0.2% |      0.13 | `EvictionProtection3Networks100Candidates`
|           76,015.27 |           13,155.25 |    0.4% |      0.85 | `EvictionProtection3Networks250Candidates`

all improvements (b1d905c225e87a4a289c0cd3593c6c21cea3fba7 p2p: earlier continuation when no remaining eviction candidates)
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           10,587.04 |           94,455.10 |    0.7% |      0.12 | `EvictionProtection0Networks250Candidates`
|            8,046.60 |          124,276.05 |    0.8% |      0.09 | `EvictionProtection1Networks250Candidates`
|           12,479.92 |           80,128.74 |    1.0% |      0.14 | `EvictionProtection2Networks250Candidates`
|            2,784.53 |          359,127.39 |    4.3% |      0.03 | `EvictionProtection3Networks050Candidates`
|            8,286.96 |          120,671.54 |    0.4% |      0.09 | `EvictionProtection3Networks100Candidates`
|           18,450.28 |           54,199.71 |    0.7% |      0.20 | `EvictionProtection3Networks250Candidates`

Benchmarks complete in ~2 seconds.

jarolrod reviewed (commented):
ACK b1d905c225e87a4a289c0cd3593c6c21cea3fba7

The way this has progressed since my last review makes sense. This is still a performance and improvement and code looks good.

I do still run into instability:
```
Unstable with ~1,001.0 iters. Increase `minEpochIterations` to e.g. 10010
```

Not sure what to make of it because I'm running the `macOS 12` Beta (not a super stable system overall). Just wanted to document that I am running into that, I will have time later today to try on other systems. If no one else is running into it, let's assume it's my system.

The warning is flaky though, here's the results of my runs where I didn't run into the warning:

**Base:** Master + (566357f8f7471f74729297868917aa32f6d3c390 + 5adb06457403f8c1d874e9c6748ecbb78ef8fa2b)


|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           18,439.40 |           54,231.71 |    4.0% |      0.21 | `EvictionProtection0Networks250Candidates`
|           16,029.50 |           62,384.99 |    3.1% |      0.18 | `EvictionProtection1Networks250Candidates`
|           43,125.31 |           23,188.24 |    2.6% |      0.47 | `EvictionProtection2Networks250Candidates`
|            7,225.39 |          138,400.78 |    3.1% |      0.08 | `EvictionProtection3Networks050Candidates`
|           15,869.58 |           63,013.65 |    3.5% |      0.18 | `EvictionProtection3Networks100Candidates`
|           97,586.25 |           10,247.35 |    1.1% |      1.06 | `EvictionProtection3Networks250Candidates`


**All Improvements:** b1d905c225e87a4a289c0cd3593c6c21cea3fba7

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           12,764.72 |           78,340.94 |    2.1% |      0.14 | `EvictionProtection0Networks250Candidates`
|            9,367.04 |          106,757.28 |    3.0% |      0.10 | `EvictionProtection1Networks250Candidates`
|           14,846.86 |           67,354.33 |    1.4% |      0.16 | `EvictionProtection2Networks250Candidates`
|            3,493.85 |          286,217.28 |    4.8% |      0.04 | `EvictionProtection3Networks050Candidates`
|           10,208.73 |           97,955.35 |    2.0% |      0.11 | `EvictionProtection3Networks100Candidates`
|           22,498.62 |           44,447.18 |    0.9% |      0.25 | `EvictionProtection3Networks250Candidates`

jonatack commented:
> Not sure what to make of it because I'm running the `macOS 12` Beta (not a super stable system overall).

Yes, I tuned my system and yet still see that most of the time on at least half of the benchmarks when running all of them with `./src/bench/bench_bitcoin`.  It came down to the preference of being able to run these new benchmarks 10 times in ~20 seconds rather than once in 30-60 seconds, and having a rapid result that works for the most people.  There just seems to be decreasing utility in adding iterations.
